src/pure_no_compaction.erl

```diff
@@ -0,0 +1,261 @@
+%%
+%% Copyright (c) 2015-2016 Christopher Meiklejohn.  All Rights Reserved.
+%% Copyright (c) 2007-2012 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Pure No Compaction CRDT: pure op-based implementation
+%% of mutate and query without POLog compaction.
+%%
+
+-module(pure_no_compaction).
+-author("Georges Younes <georges.r.younes@gmail.com>").
+
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([mutate/3, query/1]).
+
+-type pure_op() :: term().
+
+%% @doc Update with no polog compaction.
+-spec mutate(pure_op(), pure_type:id(), pure_type:crdt()) ->
+    {ok, pure_type:crdt()}.
+mutate(Op, VV, {Type, {POLog, Crystal}}) ->
+    {ok, {Type, {orddict:store(VV, Op, POLog), Crystal}}}.
+
+%% @doc Specific query for no polog compaction pure data types.
+-spec query(pure_type:crdt()) -> term().
+query({pure_gcounter, {POLog0, _Crystal}}) ->
+    orddict:fold(
+        fun(_Key, Value, Acc) ->
+            case Value of
+                increment ->
+                    Acc + 1;
+                {increment, Val} ->
+                    Acc + Val
+            end
+        end,
+        0,
+        POLog0
+    );
+query({pure_pncounter, {POLog0, _Crystal}}) ->
+    orddict:fold(
+        fun(_Key, Value, Acc) ->
+            case Value of
+                increment ->
+                    Acc + 1;
+                {increment, Val} ->
+                    Acc + Val;
+                decrement ->
+                    Acc - 1;
+                {decrement, Val} ->
+                    Acc - Val
+            end
+        end,
+        0,
+        POLog0
+    );
+query({pure_gset, {POLog0, _Crystal}}) ->
+    lists:usort(orddict:fold(
+        fun(_Key, {add, El}, Acc) ->
+                Acc ++ [El]
```

Comment:
Might be more efficient to do `[El|Acc]` and then reverse the list at the end, otherwise the `++` will be a linear cost each iteration through the loop.